### PR TITLE
release 0.14.2

### PR DESCRIPTION
**Pull request by backpacker69**

### CODE REVIEW

**Review by peerchemist — APPROVED** [review]
ok

**Review by MatthewLM — CHANGES_REQUESTED** [review]
To summarise my suggested improvement:

- Add up-to 8 inputs assuming only 1 output will be added.
- In the loop to create the split outputs
  - Re-run the split logic after determining the fee with the new required fee. Remove the complex `nDeduct` logic.
  - If the tx is over 1KB, start by removing added inputs one at a time and rerunning the loop.
- If there is only the original input left (or allowing up to 4?) and the tx still exceeds the 1KB limit, then remove as many outputs as required to bring the tx below 1KB and then split the output amount evenly across those remaining outputs.

File: src/wallet/wallet.cpp [hunk 1]
```diff
@@ -3525,18 +3526,51 @@ void CWallet::ConnectScriptPubKeyManNotifiers()
     }
 }
 
+// peercoin: function to determine optimal fraction of supply for rfc28
+double SecurityToOptimalFraction(double security, bool isTestnet) {
```
[the remark this answers is not in the document]
Comment (backpacker69) [reply src/wallet/wallet.cpp:3530]: I believe moving two functions that relate to coinstake makes no sense unless whole coinstake is split into separate file and that can be done at some later date.

File: src/wallet/wallet.cpp [hunk 2]
```diff
@@ -3525,18 +3526,51 @@ void CWallet::ConnectScriptPubKeyManNotifiers()
...
+        5.03930675692864,       // security^(1/6)
+        -4.68984257228335,      // security^(1/7)
+        1.63578550998557        // security^(1/8)
+    };
+    double optimalFraction = isTestnet ? coeffsTestnet[0] : coeffsMain[0];  // const term
+    double securityPower = security;
+
+    for (int i = 1; i < 9; ++i) {
+        optimalFraction += securityPower * (isTestnet ? coeffsTestnet[i] : coeffsMain[i]);
+        securityPower = std::pow(security, 1.0 / (i + 1));
```
[the remark this answers is not in the document]
Comment (backpacker69) [reply src/wallet/wallet.cpp:3558]: I've split it up after testnet coefficients were added as one liner became too cluttered with ?: notation

File: src/wallet/wallet.cpp [hunk 3]
```diff
@@ -3758,6 +3784,27 @@ bool CWallet::CreateCoinStake(ChainstateManager& chainman, const CWallet* pwalle
...
+    CAmount nTarget = SecurityToOptimalFraction(securityLevel, isTestnet)*supply;
+
+    if (nTarget < MIN_TARGET_OUTPUT_AMOUNT)
+        nTarget = MIN_TARGET_OUTPUT_AMOUNT;
+
+    CAmount nCombineThreshold;
+    if ((nAllowedBalance / nTarget) > maxMintingUtxos) {
+        nTarget = nAllowedBalance / maxMintingUtxos;
+        nCombineThreshold = nTarget;
+        }
```
[the remark this answers is not in the document]
Comment (backpacker69) [reply src/wallet/wallet.cpp:3804]: What's wrong with indentation?

File: src/wallet/wallet.cpp [hunk 4]
```diff
@@ -3776,27 +3823,24 @@ bool CWallet::CreateCoinStake(ChainstateManager& chainman, const CWallet* pwalle
...
             // Stop adding inputs if reached reserve limit
-            if (nCredit + pcoin->txout.nValue > nBalance - (nReserveBalance ? nReserveBalance.value() : 0))
+            if (nCredit + pcoin->txout.nValue > nAllowedBalance)
                 break;
             // Do not add additional significant input
             if (pcoin->txout.nValue > nCombineThreshold)
                 continue;
-            // Do not add input that is still too young
-            if (tx->nTime + params.nStakeMaxAge > txNew.nTime)
-                continue;
```
[the remark this answers is not in the document]
Comment (backpacker69) [reply src/wallet/wallet.cpp:3799]: They are spendable of course, but why should they be mature? Imho optimization should ignore output minting maturity and that's how I understood the rfc.

File: src/wallet/wallet.cpp [hunk 5]
```diff
@@ -3818,18 +3862,41 @@ bool CWallet::CreateCoinStake(ChainstateManager& chainman, const CWallet* pwalle
...
     {
-        // Set output amount
-        if (txNew.vout.size() == 3u+bMinterKey)
-        {
-            txNew.vout[1+bMinterKey].nValue = ((nCredit - nMinFee) / 2 / nMinFeeBase) * nMinFeeBase;
-            txNew.vout[2+bMinterKey].nValue = nCredit - nMinFee - txNew.vout[1+bMinterKey].nValue;
+        // split and set amounts based on rfc28
+        if (!bOutputsComplete) {
+            if (pwallet->m_split_coins) {
+                CAmount current = nCredit - nMinFee;
```
[the remark this answers is not in the document]
Comment (backpacker69) [reply src/wallet/wallet.cpp:3873]: That logic hasn't changed,  first run its always 0

File: src/wallet/wallet.cpp [hunk 6]
```diff
@@ -3862,9 +3929,11 @@ bool CWallet::CreateCoinStake(ChainstateManager& chainman, const CWallet* pwalle
             return error("CreateCoinStake : exceeded coinstake size limit");
 
         // Check enough fee is paid
-        if (nMinFee < GetMinFee(CTransaction(txNew), txNew.nTime) - nMinFeeBase)
+        CAmount nNeededFee = GetMinFee(CTransaction(txNew), txNew.nTime) - nMinFeeBase;
+        if (nMinFee < nNeededFee)
         {
-            nMinFee = GetMinFee(CTransaction(txNew), txNew.nTime) - nMinFeeBase;
+            nDeduct = nNeededFee - nMinFee;
```
[the remark this answers is not in the document]
Comment (backpacker69) [reply src/wallet/wallet.cpp:3935]: Perhaps we'll revisit this when adding option to specify how much user wants to spend during optimization

File: src/wallet/wallet.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (MatthewLM) [reply src/wallet/wallet.cpp:3530]: There were three functions in the RFC. The file is horrendously large and complicated already. The coinstake code should indeed be moved out. A new issue can be created for that.

File: src/wallet/wallet.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (MatthewLM) [reply src/wallet/wallet.cpp:3558]: This is confusing though as the `securityPower` input to the expression is being set after the expression. I was actually incorrect in my suggested code. Here's a new suggestion that I tested to work which also brings out the coefficients to static constants:

```
static const double coeffsMain[] = {
    -0.01205449390140,      // const
    0.00021672965052,       // security
    0.00843001158271,       // security^(1/2)
    -0.31668853152981,      // security^(1/3)
    3.15194385589535,       // security^(1/4)
    -12.93131277924088,     // security^(1/5)
    25.02314857164244,      // security^(1/6)
    -22.70985422100839,     // security^(1/7)
    7.78628320089075        // security^(1/8)
};
static const double coeffsTestnet[] = {
    -0.00277474262686,      // const
    0.00015824596035,       // security
    0.00012080445529,       // security^(1/2)
    -0.04679164943874,      // security^(1/3)
    0.56524766014644,       // security^(1/4)
    -2.50113964308348,      // security^(1/5)
    5.03930675692864,       // security^(1/6)
    -4.68984257228335,      // security^(1/7)
    1.63578550998557        // security^(1/8)
};

double SecurityToOptimalFraction(double security, bool isTestnet) {
    const auto coeffs = isTestnet ? coeffsTestnet : coeffsMain;
    double optimalFraction = coeffs[0];  // const term
    for (int i = 1; i < 9; ++i) {
        optimalFraction += std::pow(security, 1.0 / i) * coeffs[i];
    }
    return optimalFraction;
}
```

File: src/wallet/wallet.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (MatthewLM) [reply src/wallet/wallet.cpp:3804]: The `}` should have indentation reduced. That's not consistent with indentation elsewhere.

File: src/wallet/wallet.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (MatthewLM) [reply src/wallet/wallet.cpp:3799]: I meant mature for spending. If they are spendable that's OK. Just wanted to double check.

File: src/wallet/wallet.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (MatthewLM) [reply src/wallet/wallet.cpp:3873]: It only has a single run. Once `bOutputsComplete = true;` is executed, this branch will not be run again. Therefore `nMinFee` will always be 0. Once again, the branching and looping behaviour is complicated and this last part of the function needs to be reworked.

**Comment by MatthewLM** [comment]
@backpacker69 I've not tested this, but it should be clear what I'm suggesting from this code:

```
    CAmount nMinFee = 0;
    int maxOutputs = 100;
    CAmount nMinFeeBase = (IsProtocolV07(txNew.nTime) ? MIN_TX_FEE : MIN_TX_FEE_PREV7);

    while(true) {

        // Clear outputs
        txNew.vout.erase(txNew.vout.begin() + 1+bMinterKey, txNew.vout.end());

        // Assume success
        bool outputsOk = true;

        // split and set amounts based on rfc28
        if (pwallet->m_split_coins) {

            CAmount current = nCredit - nMinFee;
            double ratio = current / nTarget;

            // Obtain the optimal number of outputs and clamp it to maxOutputs to ensure the fee is not exceeded
            int desiredOutputs = min(
                int(std::floor((std::sqrt(4 * std::pow(ratio, 2) + 1) + 1) / 2)),
                maxOutputs,
            );

            if (bDebug)
                LogPrintf("rfc28: security level is %f, target amount %f, desired outputs %d from %f ppc\n",
                    securityLevel, double(nTarget)/COIN, desiredOutputs, double(current)/COIN);

            CAmount outValue = current / desiredOutputs;
            CAmount remainder = current - outValue*desiredOutputs;

            for (int i=0; i<desiredOutputs; i++) {

                // Add remainder to first output
                txNew.vout.push_back(CTxOut(outValue + (i == 0 ? remainder : 0), scriptPubKeyOut));

                // make sure transaction below 1kb
                // Why 950?
                if (::GetSerializeSize(txNew, SER_NETWORK, PROTOCOL_VERSION) > 950) {
                    outputsOk = false;
                    maxOutputs = i;
                    break;
                }

            }

        } else {
            txNew.vout.push_back(CTxOut(nCredit - nMinFee, scriptPubKeyOut));
        }

        // Sign
        // SNIPPED CODE FOR BREVITY. ADD THIS BACK

        // Limit size
        unsigned int nBytes = ::GetSerializeSize(txNew, SER_NETWORK, PROTOCOL_VERSION);
        if (nBytes >= 1000000/5)
            return error("CreateCoinStake : exceeded coinstake size limit");

        // Check enough fee is paid
        CAmount nNeededFee = GetMinFee(CTransaction(txNew), txNew.nTime) - nMinFeeBase;

        // Not enough fee paid or max outputs have changed requiring a new split and fee, so run again
        if (nMinFee < nNeededFee || !outputsOk) {
            nMinFee = nNeededFee;
            continue; // try signing again
        } else {
            if (bDebug)
                LogPrintf("CreateCoinStake : fee for coinstake %s\n", FormatMoney(nMinFee).c_str());
            break;
        }

    }
```

File: src/wallet/wallet.cpp [hunk 7]
```diff
@@ -3862,41 +3860,49 @@ bool CWallet::CreateCoinStake(ChainstateManager& chainman, const CWallet* pwalle
...
+                    securityLevel, double(nTargetOutputAmount)/COIN, desiredOutputs, double(current)/COIN);
+
+            CAmount outValue = current / desiredOutputs;
+            CAmount remainder = current - outValue*desiredOutputs;
+
+            for (int i=0; i<desiredOutputs; i++) {
+                // Add remainder to first output
+                txNew.vout.push_back(CTxOut(outValue + (i == 0 ? remainder : 0), scriptPubKeyOut));
+                // make sure transaction below 1kb
+                if (::GetSerializeSize(txNew, SER_NETWORK, PROTOCOL_VERSION) > 1024) {
```
Comment (MatthewLM) [inline src/wallet/wallet.cpp:3894]: Isn't the fee-free amount 1000 bytes?

**Review by MatthewLM — COMMENTED** [review]
Looks good, though please check the fee-free coinstake size.

I've not tested it however. Certain edge cases should be tested before release (eg. when fee calculation causes a drop in the desired outputs). Automated functional tests would be ideal, but without that manual testing of the different branching behaviour and constraints should be done.

I don't expect there should be any problems but I would release this as an RC and ask people to test out on testnet to see if any problems arise before final release.

File: src/wallet/wallet.cpp [hunk 8]
```diff
@@ -3776,27 +3821,24 @@ bool CWallet::CreateCoinStake(ChainstateManager& chainman, const CWallet* pwalle
...
         {
-            // Stop adding more inputs if already too many inputs
-            if (txNew.vin.size() >= 100)
+            // Stop adding more inputs if already too many inputs and we are above target
+            if ((txNew.vin.size() >= MAX_COINSTAKE_INPUTS) && (nCredit > nTargetOutputAmount))
                 break;
-            // Stop adding more inputs if value is already pretty significant
-            if (nCredit > nCombineThreshold)
+            // Stop adding more inputs if we are below target but about to exceed 1kb
+            if ((txNew.vin.size() >= 8) && (nCredit < nTargetOutputAmount))
```
Comment (MatthewLM) [inline src/wallet/wallet.cpp:3833]: Do 8 maximally sized inputs fit when there is the minter key output in addition to the other output?

File: src/wallet/wallet.cpp (hunk 7 above)
Comment (MatthewLM) [reply src/wallet/wallet.cpp:3894]: Also I think we should add an extra condition that `i` must be greater than 0. That shouldn't happen but it protects against the possibility that `maxOutputs` would get set to 0.

File: src/wallet/wallet.cpp [hunk 9]
```diff
@@ -3818,17 +3860,48 @@ bool CWallet::CreateCoinStake(ChainstateManager& chainman, const CWallet* pwalle
     }
 
     CAmount nMinFee = 0;
+    int maxOutputs = 30;
     CAmount nMinFeeBase = (IsProtocolV07(txNew.nTime) ? MIN_TX_FEE : MIN_TX_FEE_PREV7);
```
[the remark this answers is not in the document]
Comment (MatthewLM) [reply src/wallet/wallet.cpp:3864]: As mentioned, the `nMinFeeBase` should be `PERKB_TX_FEE` since that's what is available.

File: src/wallet/wallet.cpp (hunk 8 above)
Comment (backpacker69) [reply src/wallet/wallet.cpp:3833]: good catch about minter key, i've added the condition during adding inputs

**Comment by backpacker69** [comment]
i have manually tested the edge cases including minter key and added extra safety handling of situation when size of inputs exceeds 1000 bytes.

File: src/wallet/wallet.cpp [hunk 10]
```diff
@@ -3891,7 +3891,9 @@ bool CWallet::CreateCoinStake(ChainstateManager& chainman, const CWallet* pwalle
                 // Add remainder to first output
                 txNew.vout.push_back(CTxOut(outValue + (i == 0 ? remainder : 0), scriptPubKeyOut));
                 // make sure transaction below 1kb
-                if (::GetSerializeSize(txNew, SER_NETWORK, PROTOCOL_VERSION) > 1024) {
+                if (::GetSerializeSize(txNew, SER_NETWORK, PROTOCOL_VERSION) > 1000) {
+                    if (!i)
+                        return error("CreateCoinStake : failed to create coinstake");
```
[the remark this answers is not in the document]
Comment (backpacker69) [reply src/wallet/wallet.cpp:3896]: until we bring in options to allow exceeding 1000 byte limit, i don't want to create any coinstakes over the limit, i don't think it's right, as we don't know by how much that limit has been exceeded.

[the remark this answers is not in the document]
Comment (MatthewLM) [reply src/wallet/wallet.cpp:3896]: It shouldn't get exceeded if all is well with the code as it is now. I was just thinking about having code to ensure the coinstake is still created even if the code is to change in the future.

However, it should be fine if you'd like to resolve this and merge.

I think it would be a good idea to do a release candidate so it can be tried on testnet (or mainnet for those willing to try it there) for a little while before final release.

**Review by MatthewLM — APPROVED** [review]
